core/src/main/java/io/grpc/CallOptions.java

```diff
@@ -197,47 +218,55 @@ public String getCompressor() {
   @Nullable
   @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1767")
   public String getAuthority() {
-    return authority;
+    return getOption(AUTHORITY_KEY);
   }
 
   /**
    * Returns a new {@code CallOptions} with {@code executor} to be used instead of the default
    * executor specified with {@link ManagedChannelBuilder#executor}.
    */
   public CallOptions withExecutor(Executor executor) {
-    CallOptions newOptions = new CallOptions(this);
-    newOptions.executor = executor;
-    return newOptions;
+    return withOption(EXECUTOR_KEY, executor);
   }
 
   @Nullable
   public Executor getExecutor() {
-    return executor;
+    return getOption(EXECUTOR_KEY);
   }
 
+  /**
+   * Create a copy of this CallOptions object with key set to value.
+   */
+  public <T> CallOptions withOption(Key<T> key, T value) {
+    CallOptions newOptions = new CallOptions(this);
```

Comment:
Yes, I am concerned with the performance of copying the hash map. I'm okay with copying it only when `withOption` is called, since it will be rarer and we can use something quick and easy now and only improve its performance when necessary.